internal/gps/source_cache_bolt_encode.go

```diff
@@ -5,140 +5,94 @@
 package gps
 
 import (
-	"bytes"
 	"encoding/binary"
-	"fmt"
-	"strings"
 	"time"
 
 	"github.com/boltdb/bolt"
+	"github.com/golang/dep/internal/gps/internal/pb"
 	"github.com/golang/dep/internal/gps/pkgtree"
+	"github.com/golang/protobuf/proto"
+	"github.com/jmank88/nuts"
 	"github.com/pkg/errors"
 )
 
-// cacheEncodeUnpairedVersion returns an encoded UnpairedVersion.
-func cacheEncodeUnpairedVersion(uv UnpairedVersion) ([]byte, error) {
-	var pre string
-	switch uv.Type() {
-	case IsBranch:
-		if uv.(branchVersion).isDefault {
-			pre = "defaultBranch:"
-		} else {
-			pre = "branch:"
-		}
-	case IsSemver, IsVersion:
-		pre = "ver:"
-	default:
-		return nil, fmt.Errorf("unrecognized version type: %d", uv.Type())
-	}
-	return []byte(pre + uv.String()), nil
-}
-
-// cacheDecodeUnpairedVersion decodes and returns a new UnpairedVersion.
-func cacheDecodeUnpairedVersion(b []byte) (UnpairedVersion, error) {
-	const br, dbr, ver = "branch:", "defaultBranch:", "ver:"
-	s := string(b)
-	switch {
-	case strings.HasPrefix(s, br):
-		return NewBranch(strings.TrimPrefix(s, br)), nil
-	case strings.HasPrefix(s, dbr):
-		return newDefaultBranch(strings.TrimPrefix(s, dbr)), nil
-	case strings.HasPrefix(s, ver):
-		return NewVersion(strings.TrimPrefix(s, ver)), nil
-	default:
-		return nil, fmt.Errorf("unrecognized prefix: %s", s)
-	}
-}
+var (
+	cacheKeyC = []byte("c")
```

Comment:
They don't map 1-1, which is why I opted to name them what they are, not how they are used.

Is this better?

```go
var (
	cacheKeyConstraint = []byte("c")
	cacheKeyComment = cacheKeyConstraint
	cacheKeyError = []byte("e")
	cacheKeyHash = []byte("h")
	cacheKeyIgnored = []byte("i")
	cacheKeyImport = cacheKeyIgnored
	cacheKeyLock = []byte("l")
	cacheKeyName = []byte("n")
	cacheKeyOverride = []byte("o")
	cacheKeyPTree = []byte("p")
	cacheKeyRequired = []byte("r")
	cacheKeyRevision = cacheKeyRequired
	cacheKeyTestImport = []byte("t")

	cacheRevision = byte('r')
	cacheVersion = byte('v')
)
```

I wish this all could have dropped into a separate cache package to avoid all the annoying prefixes.

Reply:
(Regarding splitting packages, IIRC it would require _at least_ (maybe only?) pushing version stuff into a separate `gps/version` package to avoid cycles)

Reply:
ah yeah, i see. in this case, though, a bit of duplication for the reader's sake is acceptable.

> (Regarding splitting packages, IIRC it would require at least (maybe only?) pushing version stuff into a separate gps/version package to avoid cycles)

i've considered such a thing quite a bit, but i don't think we can currently manage it because of all the reliance on hidden properties. i'd be open to refactoring that and just exporting everything - i was probably defending against a bogeyman by not exporting it - but i don't want to make that a yak we have to shave for this.

Reply:
I started a separate (from `master`) experimental conversion, leveraging aliases quite heavily for now to minimize churn, and it's not actually too bad - plus some of the API ends up cleaner.  I dropped in a few exported helper functions, rather than exporting all the internals.  Not that helper functions are necessarily a better solution than exporting the types, but it proved the concept without too much refactoring.

This can all be considered separately from this PR and caching in general though.